File: smoke_tests/ruby/sensitive_data.rb
# Sensitive Data Exposure vulnerabilities in Ruby/Rails
class SensitiveDataController < ApplicationController
  # Test 1: Exception details exposed
  def show_error
    begin
      raise_error
    rescue => e
      # VULNERABLE: Stack trace exposed
      render plain: "Error: #{e.message}\n#{e.backtrace.join("\n")}"
    end
  end

  # Test 2: Debug info in response
  def get_data
    data = load_data
    # VULNERABLE: Debug information exposed
    render json: {
      data: data,
      debug: {
        server: Socket.gethostname,
        user: ENV['USER'],
        rails_env: Rails.env
      }
    }
  end

  # Test 3: Logging sensitive data
  def process_payment
    card_number = params[:card_number]
    cvv = params[:cvv]
    # VULNERABLE: Credit card data logged
    Rails.logger.info "Processing card: #{card_number}, CVV: #{cvv}"
    process_card(card_number, cvv)
    head :ok
  end

  # Test 4: Sensitive data in URL
  def show_account
    ssn = params[:ssn]
    account = params[:account_number]
    # VULNERABLE: SSN and account in URL (logged, cached)
    @user = User.find_by(ssn: ssn)
    render :show
  end

  # Test 5: Caching sensitive responses
  def get_user_details
    # VULNERABLE: Sensitive data being cached
    expires_in 1.hour
    render json: {
      user_id: current_user.id,
      email: current_user.email,
      ssn: current_user.ssn
    }
  end

  # Test 6: Unencrypted sensitive storage
  def store_ssn
    ssn = params[:ssn]
    # VULNERABLE: SSN stored unencrypted
    File.write('/data/user.txt', ssn)
    head :ok
  end

  # Test 7: API key in response
  def get_config
    # VULNERABLE: API keys exposed
    render json: {
      api_endpoint: 'https://api.example.com',
      api_key: 'sk-12345-secret-key',
      db_password: 'secret123'
    }
  end

  # Test 8: Missing security headers
  def secure_page
    # VULNERABLE: No security headers set
    # Missing: X-Content-Type-Options, X-Frame-Options, CSP
    render :secure
  end

  # Test 9: HTTP for sensitive data
  def redirect_to_payment
    # VULNERABLE: HTTP for payment page
    redirect_to 'http://payment.example.com/checkout'
  end

  # Test 10: Instance variables with PII
  def profile
    @ssn = current_user.ssn
    @credit_card = current_user.credit_card
    # VULNERABLE: PII in instance variables exposed to view
    render :profile
  end

  # Test 11: Verbose errors in JSON API
  def api_error
    begin
      process_api_request
    rescue => e
      # VULNERABLE: Full error details in API response
      render json: {
        error: e.message,
        backtrace: e.backtrace,
        class: e.class.name
      }, status: 500
    end
  end

  # Test 12: Secrets in view
  def admin_panel
    @database_url = ENV['DATABASE_URL']
    @secret_key = Rails.application.secrets.secret_key_base
    # VULNERABLE: Secrets passed to view
    render :admin_panel
  end

  private

  def raise_error
    raise StandardError, 'Something went wrong'
  end

  def load_data
    {}
  end

  def process_card(number, cvv); end
  def process_api_request; end
end

File: smoke_tests/javascript/sensitive_data.js
// Sensitive Data Exposure vulnerabilities in JavaScript/Node.js
const fs = require('fs');

// Test 1: Exception details exposed
function showError(req, res) {
    try {
        throw new Error('Database connection failed');
    } catch (e) {
        // VULNERABLE: Stack trace exposed
        res.status(500).json({
            error: e.message,
            stack: e.stack
        });
    }
}

// Test 2: Debug info in response
function getData(req, res) {
    const data = loadData();
    // VULNERABLE: Debug information exposed
    res.json({
        data,
        debug: {
            server: process.env.HOSTNAME,
            nodeVersion: process.version,
            cwd: process.cwd()
        }
    });
}

// Test 3: Logging sensitive data
function processPayment(req, res) {
    const { cardNumber, cvv } = req.body;
    // VULNERABLE: Credit card data logged
    console.log(`Processing card: ${cardNumber}, CVV: ${cvv}`);
    doPayment(cardNumber, cvv);
    res.json({ success: true });
}

// Test 4: Sensitive data in URL
function showAccount(req, res) {
    const { ssn, accountNumber } = req.query;
    // VULNERABLE: SSN and account in URL (logged, cached)
    const user = getUserBySsn(ssn);
    res.json(user);
}

// Test 5: Caching sensitive responses
function getUserDetails(req, res) {
    // VULNERABLE: Sensitive data being cached
    res.set('Cache-Control', 'public, max-age=3600');
    res.json({
        userId: req.session.userId,
        email: getUserEmail(),
        ssn: getUserSsn()
    });
}

// Test 6: Unencrypted sensitive storage
function storeSsn(req, res) {
    const { ssn } = req.body;
    // VULNERABLE: SSN stored unencrypted
    fs.writeFileSync('/data/user.txt', ssn);
    res.json({ success: true });
}

// Test 7: API key in response
function getConfig(req, res) {
    // VULNERABLE: API keys exposed
    res.json({
        apiEndpoint: 'https://api.example.com',
        apiKey: 'sk-12345-secret-key',
        dbPassword: 'secret123'
    });
}

// Test 8: Missing security headers
function securePage(req, res) {
    // VULNERABLE: No security headers
    // Missing: X-Content-Type-Options, X-Frame-Options, CSP
    res.send('<html>Content</html>');
}

// Test 9: HTTP for sensitive data
function redirectToPayment(req, res) {
    // VULNERABLE: HTTP for payment page
    res.redirect('http://payment.example.com/checkout');
}

// Test 10: Environment variables exposed
function envInfo(req, res) {
    // VULNERABLE: Environment variables exposed
    res.json({
        env: process.env
    });
}

// Test 11: Source maps in production
// webpack.config.js with devtool: 'source-map' in production
// VULNERABLE: Source code exposed via source maps

// Test 12: Error handler with details
function errorHandler(err, req, res, next) {
    // VULNERABLE: Full error details in production
    res.status(500).json({
        message: err.message,
        stack: err.stack,
        code: err.code
    });
}

// Test 13: Verbose GraphQL errors
const graphqlConfig = {
    // VULNERABLE: Detailed errors in production
    formatError: (error) => ({
        message: error.message,
        locations: error.locations,
        path: error.path,
        stack: error.stack
    })
};

// Test 14: PII in logs
function logRequest(req, res, next) {
    // VULNERABLE: PII logged
    console.log(`Request from ${req.ip}: ${JSON.stringify(req.body)}`);
    next();
}

// Helper functions
function loadData() { return {}; }
function doPayment(card, cvv) {}
function getUserBySsn(ssn) { return {}; }
function getUserEmail() { return 'user@example.com'; }
function getUserSsn() { return '123-45-6789'; }

module.exports = {
    showError,
    getData,
    processPayment,
    showAccount,
    getUserDetails,
    storeSsn,
    getConfig,
    securePage,
    redirectToPayment,
    envInfo,
    errorHandler,
    graphqlConfig,
    logRequest
};
